template/template.go

```diff
@@ -296,3 +312,196 @@ func (e *executer) resolveOptions(options []ast.Option) (map[string]any, error)
 
 	return m, nil
 }
+
+// https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#pattern-selection
+func (e *executer) resolveMatcher(m ast.Matcher) (string, error) {
+	output, err := e.patternSelection(m)
+	if err != nil {
+		return "", err
+	}
+
+	return output.String(), nil
+}
+
+func (e *executer) patternSelection(m ast.Matcher) (strings.Builder, error) {
+	res, err := e.resolveSelector()
+	if err != nil {
+		return strings.Builder{}, err
+	}
+
+	pref := e.resolvePreferences(m, res)
+
+	filteredVariants := e.filterVariants(m, pref)
+
+	sortable := e.sortVariants(filteredVariants, pref)
+
+	output, err := e.selectBestVariant(sortable)
+	if err != nil {
+		return strings.Builder{}, err
+	}
+
+	return output, nil
+}
+
+func (e *executer) resolveSelector() ([]Selector, error) {
+	// Step 1: Resolve Selector Value (Modified)
+	var res []Selector
+
+	for k, v := range e.input {
```

Comment:
> For each selector sel, in source order

so the loop should be across matcher expressions, not the input map. The whole idea seems wrong to me :)
Here is my draft, which should be correct

```go
func (e *executer) resolveSelector2(matcher ast.Matcher) ([]any, error) {
	res := make([]any, 0, len(matcher.MatchStatements))

	for _, selectExpr := range matcher.MatchStatements {
		var function ast.Function

		switch annotation := selectExpr.Annotation.(type) {
		case nil:
			// error: Select Expression cannot be without annotation
		case ast.ReservedAnnotation, ast.PrivateUseAnnotation:
			// error: Select Expression cannot be with reserved or private use annotation
		case ast.Function:
			function = annotation
		}

		registryF, ok := e.template.funcRegistry[function.Identifier.Name]
		if !ok {
			// error: Unknown function reference
		}

		opts, err := e.resolveOptions(function.Options)
		if err != nil {
			// error: Resolve Options
		}

		rslt, err := registryF.Match(selectExpr.Operand, opts)
		if err != nil {
			// error: Selection Error
		}

		res = append(res, rslt)

	}

	return res, nil
}
```